File: cmd/keychron.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"
)

const (
	VID              = 0x3434
	PID              = 0x0370
	RawHIDInterface  = 1
	RawHIDBufferSize = 32

	ViaSet  = 0x07
	ViaGet  = 0x08
	ViaSave = 0x09

	ValBrightness = 0x80
	ValEffect     = 0x81
	ValSpeed      = 0x82
	ValColor      = 0x83 // 2 bytes: hue, saturation
)

// KeychronV7 controls the keyboard RGB LEDs via raw HID over /dev/hidraw.
type KeychronV7 struct {
	file *os.File
}

// findHidrawDevice scans sysfs for the Keychron V7 raw HID interface.
func findHidrawDevice() (string, error) {
	matches, err := filepath.Glob("/sys/class/hidraw/hidraw*")
	if err != nil {
		return "", err
	}

	hidID := fmt.Sprintf("0003:%08X:%08X", VID, PID)

	for _, entry := range matches {
		uevent, err := os.ReadFile(filepath.Join(entry, "device", "uevent"))
		if err != nil {
			continue
		}

		found := false
		for _, line := range strings.Split(string(uevent), "\n") {
			if strings.TrimSpace(line) == "HID_ID="+hidID {
				found = true
				break
			}
		}
		if !found {
			continue
		}

		// Resolve the device symlink to find the USB interface directory.
		devicePath, err := filepath.EvalSymlinks(filepath.Join(entry, "device"))
		if err != nil {
			continue
		}

		// Walk up to find bInterfaceNumber.
		// The device path is like .../usb1/1-3/1-3:1.1/0003:3434:0370.xxxx
		// The USB interface dir (with bInterfaceNumber) is the parent.
		ifDir := filepath.Dir(devicePath)
		ifNum, err := os.ReadFile(filepath.Join(ifDir, "bInterfaceNumber"))
		if err != nil {
			continue
		}

		if strings.TrimSpace(string(ifNum)) == fmt.Sprintf("%02d", RawHIDInterface) {
			devName := filepath.Base(entry)
			return "/dev/" + devName, nil
		}
	}

	return "", fmt.Errorf("Keychron V7 not found. Is it plugged in? Are udev rules installed?")
}

func (kb *KeychronV7) Open() error {
	path, err := findHidrawDevice()
	if err != nil {
		return err
	}
	f, err := os.OpenFile(path, os.O_RDWR, 0)
	if err != nil {
		return err
	}
	kb.file = f
	return nil
}

func (kb *KeychronV7) Close() {
	if kb.file != nil {
		kb.file.Close()
		kb.file = nil
	}
}

func (kb *KeychronV7) send(data []byte) ([]byte, error) {
	buf := make([]byte, RawHIDBufferSize)
	copy(buf, data)

	_, err := kb.file.Write(buf)
	if err != nil {
		return nil, fmt.Errorf("hid write: %w", err)
	}

	time.Sleep(20 * time.Millisecond)

	kb.file.SetReadDeadline(time.Now().Add(200 * time.Millisecond))
	resp := make([]byte, RawHIDBufferSize)
	n, err := kb.file.Read(resp)
	if err != nil {
		// Timeout is not fatal — some commands don't respond.
		if os.IsTimeout(err) {
			return nil, nil
		}
		return nil, fmt.Errorf("hid read: %w", err)
	}
	return resp[:n], nil
}

func (kb *KeychronV7) setValue(valueID byte, data ...byte) {
	msg := []byte{ViaSet, valueID}
	msg = append(msg, data...)
	kb.send(msg)
}

func (kb *KeychronV7) getValue(valueID byte, numBytes int) []byte {
	resp, err := kb.send([]byte{ViaGet, valueID})
	if err != nil || resp == nil || len(resp) < 2+numBytes {
		return nil
	}
	return resp[2 : 2+numBytes]
}

func (kb *KeychronV7) Save() {
	kb.send([]byte{ViaSave})
}

func (kb *KeychronV7) SetEffect(effect string) {
	if idx, ok := EffectByName(effect); ok {
		kb.setValue(ValEffect, byte(idx))
	} else {
		// Default to solid if unknown.
		kb.setValue(ValEffect, 1)
	}
}

func (kb *KeychronV7) SetEffectNum(effect int) {
	kb.setValue(ValEffect, byte(effect))
}

func (kb *KeychronV7) SetSpeed(speed int) {
	kb.setValue(ValSpeed, byte(clamp(speed, 0, 3)))
}

func (kb *KeychronV7) SetBrightness(val int) {
	kb.setValue(ValBrightness, byte(clamp(val, 0, 255)))
}

func (kb *KeychronV7) SetColor(hue, sat int) {
	kb.setValue(ValColor, byte(clamp(hue, 0, 255)), byte(clamp(sat, 0, 255)))
}

func (kb *KeychronV7) SetColorHSV(hue, sat, brightness int) {
	kb.SetColor(hue, sat)
	kb.SetBrightness(brightness)
}

func (kb *KeychronV7) GetState() map[string]interface{} {
	state := map[string]interface{}{
		"brightness": nil,
		"effect":     nil,
		"speed":      nil,
		"hue":        nil,
		"saturation": nil,
	}

	if v := kb.getValue(ValBrightness, 1); v != nil {
		state["brightness"] = int(v[0])
	}
	if v := kb.getValue(ValEffect, 1); v != nil {
		state["effect"] = int(v[0])
	}
	if v := kb.getValue(ValSpeed, 1); v != nil {
		state["speed"] = int(v[0])
	}
	if v := kb.getValue(ValColor, 2); v != nil {
		state["hue"] = int(v[0])
		state["saturation"] = int(v[1])
	}

	return state
}

// Animations

func (kb *KeychronV7) CycleColors(duration float64) {
	kb.SetEffect("solid")
	kb.SetColor(0, 255)
	kb.SetBrightness(255)
	start := time.Now()
	hue := 0.0
	for time.Since(start).Seconds() < duration {
		kb.SetColor(int(hue*255)%256, 255)
		hue = hue + 0.01
		if hue >= 1.0 {
			hue -= 1.0
		}
		time.Sleep(50 * time.Millisecond)
	}
}

func (kb *KeychronV7) PoliceLights(duration float64) {
	kb.SetEffect("solid")
	kb.SetBrightness(255)
	start := time.Now()
	for time.Since(start).Seconds() < duration {
		kb.SetColor(0, 255) // red
		time.Sleep(120 * time.Millisecond)
		kb.SetBrightness(0)
		time.Sleep(60 * time.Millisecond)
		kb.SetBrightness(255)
		kb.SetColor(170, 255) // blue
		time.Sleep(120 * time.Millisecond)
		kb.SetBrightness(0)
		time.Sleep(60 * time.Millisecond)
		kb.SetBrightness(255)
	}
}

func (kb *KeychronV7) Pulse(hue int, duration float64) {
	kb.SetEffect("solid")
	kb.SetColor(hue, 255)
	start := time.Now()
	for time.Since(start).Seconds() < duration {
		for b := 0; b < 255; b += 5 {
			if time.Since(start).Seconds() >= duration {
				return
			}
			kb.SetBrightness(b)
			time.Sleep(30 * time.Millisecond)
		}
		for b := 255; b > 0; b -= 5 {
			if time.Since(start).Seconds() >= duration {
				return
			}
			kb.SetBrightness(b)
			time.Sleep(30 * time.Millisecond)
		}
	}
}

func (kb *KeychronV7) FlashAlert(hue int, duration float64) {
	kb.SetEffect("solid")
	kb.SetColor(hue, 255)
	start := time.Now()
	for time.Since(start).Seconds() < duration {
		kb.SetBrightness(255)
		time.Sleep(150 * time.Millisecond)
		kb.SetBrightness(0)
		time.Sleep(100 * time.Millisecond)
	}
}

func clamp(val, min, max int) int {
	if val < min {
		return min
	}
	if val > max {
		return max
	}
	return val
}

File: saengsation/keychron.py
"""Keychron V7 RGB control via QMK VIA raw HID protocol.

VIA protocol v10 (0x0A) uses flat lighting commands with no channel byte.
Format: [command_id, value_id, data...]

Value IDs (confirmed by hardware probing):
  0x80 = brightness (0-255)
  0x81 = effect/mode
  0x82 = speed (0-3)
  0x83 = color (hue, saturation as 2 bytes)
"""

import hid
import time

VID = 0x3434
PID = 0x0370
RAW_HID_INTERFACE = 1
RAW_HID_BUFFER_SIZE = 32

# VIA protocol command IDs
VIA_SET = 0x07
VIA_GET = 0x08
VIA_SAVE = 0x09

# VIA v10 lighting value IDs
VAL_BRIGHTNESS = 0x80
VAL_EFFECT = 0x81
VAL_SPEED = 0x82
VAL_COLOR = 0x83  # 2 bytes: hue, saturation

# Known QMK RGB Matrix effects (common subset, firmware may support fewer)
EFFECTS = {
    "off": 0,
    "solid": 1,
    "breathing": 2,
    "band_spiral_val": 3,
    "cycle_all": 4,
    "cycle_left_right": 5,
    "cycle_up_down": 6,
    "rainbow_moving_chevron": 7,
    "cycle_out_in": 8,
    "cycle_out_in_dual": 9,
    "cycle_pinwheel": 10,
    "cycle_spiral": 11,
    "dual_beacon": 12,
    "rainbow_beacon": 13,
    "jellybean_raindrops": 14,
    "pixel_rain": 15,
    "typing_heatmap": 16,
    "digital_rain": 17,
    "reactive_simple": 18,
    "reactive_multiwide": 19,
    "reactive_multinexus": 20,
    "splash": 21,
    "solid_splash": 22,
}


class KeychronV7:
    def __init__(self):
        self._dev = None

    def open(self):
        self._dev = hid.device()
        path = self._find_raw_hid_path()
        if path is None:
            raise RuntimeError(
                "Keychron V7 not found. Is it plugged in? Are udev rules installed?"
            )
        self._dev.open_path(path)
        self._dev.set_nonblocking(True)

    def close(self):
        if self._dev:
            self._dev.close()
            self._dev = None

    def __enter__(self):
        self.open()
        return self

    def __exit__(self, *args):
        self.close()

    def _find_raw_hid_path(self) -> bytes | None:
        for dev in hid.enumerate(VID, PID):
            if dev["interface_number"] == RAW_HID_INTERFACE:
                return dev["path"]
        return None

    def _send(self, data: list[int]) -> list[int]:
        buf = [0x00] + data + [0x00] * (RAW_HID_BUFFER_SIZE - len(data))
        self._dev.write(buf)
        time.sleep(0.02)
        resp = self._dev.read(RAW_HID_BUFFER_SIZE, timeout_ms=200)
        return resp if resp else []

    def _set_value(self, value_id: int, *data: int):
        self._send([VIA_SET, value_id, *data])

    def _get_value(self, value_id: int, num_bytes: int = 1) -> list[int] | int | None:
        resp = self._send([VIA_GET, value_id])
        if resp and len(resp) >= 2 + num_bytes:
            if num_bytes == 1:
                return resp[2]
            return list(resp[2:2 + num_bytes])
        return None

    def save(self):
        self._send([VIA_SAVE])

    def set_effect(self, effect: int | str):
        if isinstance(effect, str):
            effect = EFFECTS.get(effect.lower(), 1)
        self._set_value(VAL_EFFECT, effect)

    def set_speed(self, speed: int):
        self._set_value(VAL_SPEED, max(0, min(3, speed)))

    def set_brightness(self, val: int):
        self._set_value(VAL_BRIGHTNESS, max(0, min(255, val)))

    def set_color(self, hue: int, sat: int):
        self._set_value(VAL_COLOR, max(0, min(255, hue)), max(0, min(255, sat)))

    def set_hue(self, hue: int):
        color = self._get_value(VAL_COLOR, 2)
        sat = color[1] if color else 255
        self.set_color(hue, sat)

    def set_saturation(self, sat: int):
        color = self._get_value(VAL_COLOR, 2)
        hue = color[0] if color else 0
        self.set_color(hue, sat)

    def set_color_hsv(self, hue: int, sat: int, brightness: int):
        self.set_color(hue, sat)
        self.set_brightness(brightness)

    def get_state(self) -> dict:
        color = self._get_value(VAL_COLOR, 2)
        return {
            "brightness": self._get_value(VAL_BRIGHTNESS),
            "effect": self._get_value(VAL_EFFECT),
            "speed": self._get_value(VAL_SPEED),
            "hue": color[0] if color else None,
            "saturation": color[1] if color else None,
        }

    def cycle_colors(self, duration: float = 10.0, step_delay: float = 0.05):
        self.set_effect("solid")
        self.set_color(0, 255)
        self.set_brightness(255)
        start = time.time()
        hue = 0
        while time.time() - start < duration:
            self.set_color(hue % 256, 255)
            hue += 3
            time.sleep(step_delay)

    def pulse(self, hue: int = 0, cycles: int = 5, step_delay: float = 0.03):
        self.set_effect("solid")
        self.set_color(hue, 255)
        for _ in range(cycles):
            for brightness in range(0, 255, 5):
                self.set_brightness(brightness)
                time.sleep(step_delay)
            for brightness in range(255, 0, -5):
                self.set_brightness(brightness)
                time.sleep(step_delay)

    def flash_alert(self, hue: int = 0, flashes: int = 5, on_time: float = 0.15, off_time: float = 0.1):
        self.set_effect("solid")
        self.set_color(hue, 255)
        for _ in range(flashes):
            self.set_brightness(255)
            time.sleep(on_time)
            self.set_brightness(0)
            time.sleep(off_time)

    def police_lights(self, duration: float = 10.0, flash_time: float = 0.12):
        self.set_effect("solid")
        self.set_brightness(255)
        start = time.time()
        while time.time() - start < duration:
            self.set_color(0, 255)  # red
            time.sleep(flash_time)
            self.set_brightness(0)
            time.sleep(flash_time * 0.5)
            self.set_brightness(255)
            self.set_color(170, 255)  # blue
            time.sleep(flash_time)
            self.set_brightness(0)
            time.sleep(flash_time * 0.5)
            self.set_brightness(255)
